v2/devdoc/logging.md

```diff
@@ -0,0 +1,339 @@
+# `c-logging` requirements
+
+## Basic functionality
+
+`c-logging` shall expose APIs allowing the user to log events that can be used for observing the software activity, debugging, etc.
+
+The exposed API shall allow creating contexts both dynamically and on the stack and using those contexts for logging.
+
+The API shall allow easy formatting of values to be placed in the contexts by using printf-like formatting.
+
+```
+void logger_log(LOG_LEVEL log_level, LOG_CONTEXT_HANDLE log_context, const char* file, const char* func, int line_no, const char* format, ...);
+
+#define LOGGER_LOG(log_level, log_context, format, ...) \
+    logger_log(log_level, log_context, __FILE__, __FUNCTION__, __LINE__, format MU_IFCOMMALOGIC(MU_COUNT_ARG(__VA_ARGS__)) __VA_ARGS__)
+
+#define LOGGER_LOG_EX(log_level, ...) \
+    ...
+
+// macro that can be used to create a dynamically allocated context
+#define LOG_CONTEXT_CREATE(dest_log_context, parent_context, ...) \
+    ...
+
+// destroy a dynamically allocated context
+#define LOG_CONTEXT_DESTROY(log_context_handle) \
+    ...
+
+// macro that can be used to create a context on the stack.
+// stack allocated contexts do not require any destructor
+#define LOG_CONTEXT_LOCAL_DEFINE(log_context, parent_context, ...) \
+    ...
+
+// when a user wants to specify the log context type name, the user can use this macro
+// otherwise, the default is to use the context variable name as type name
+// LOG_CONTEXT_TYPE_NAME(log_context_type_name)
+
+// when a user wants to specify properties for a context, they can use LOG_CONTEXT_PROPERTY
```

Comment:
>LOG_CONTEXT_PROPERTY [](http://example.com/codeflow?start=71&length=20)

Why doesn't this need a "property_conversion_specfier" like line 41? How does it know what to use?

Reply:
It uses the format specifier predefined for the given basic type (int32_t would use PRId32, etc.).